packages/server/src/workers/download.ts

```diff
@@ -253,28 +268,54 @@ export async function execDownloadJob<T extends Resource = Resource>(job: Job<Do
     // Note that while the Fetch Standard requires the property to always be a WHATWG ReadableStream, in node-fetch it is a Node.js Readable stream.
     await getBinaryStorage().writeBinary(binary, contentDisposition, contentType, response.body as Readable);
 
-    const updated = JSON.parse(JSON.stringify(resource).replace(url, `Binary/${binary.id}`)) as Resource;
-    (updated.meta as Meta).author = { reference: 'system' };
-    await systemRepo.updateResource(updated);
+    // re-fetch resource so we are mutating as recent a copy as possible
+    // (there may have been other mutations applied while we were writing the
+    // object into storage)
+    resource = await systemRepo.readResource<T>(resourceType, id);
+
+    const attachments = getAttachments(resource);
+    const patches = attachments
+      .filter(attachment => attachment.value.url === url)
+      .map(value => ({
+        op: 'replace' as const,
+        path: `${pathToPointer(value.path)}/url`,
+        value: `Binary/${binary.id}`,
+      }));
```

Comment:
I think we should bail out if `patches.length === 0`. This would mean that since we started this job someone else has already removed the URL we fetched from the resource. We might even want to destroy the item we just wrote into storage since it won't have any pointers to it.